docs: add documentation for Generator

samuel-duhaime opened the pull request:
docs: add documentation for generator

Comments: I'm starting to export the Generator survey project to Evolution, so we can use it in all the projects. I'm going to push the others scripts (soon) in others pull request, so we can review them more easily.

tahini commented on generator/generator.md:
```diff
@@ -0,0 +1,246 @@
+To run this script, follow these steps:
+
+1. Create an Excel file following this example : [Enquete_Teletravail_Generate_survey](https://polymtlca0-my.sharepoint.com/:x:/g/personal/samuel_duhaime-morissette_polymtl_ca/Ecp5FW8DQ5JFiYDEMCOb9PMBSUUbxdsaFq4_ncT5oxchWw?e=jsdIJt).
+2. Update the `.env` file with the necessary environment variables.
```
Download the file should be separate from the generation. Ideally, this step would be optional and the survey can be generated from a local file

samuel-duhaime replied:
Yes, I have a YAML file to activate or not the generateExcel.py. So we can already, generate the survey in local, or if the user prefer with the Excel Cloud. 

```YAML
excel:
    active_script: true
```

tahini commented on generator/generator.md:
````diff
@@ -0,0 +1,246 @@
+    OFFICE365_PASSWORD = "<your_office365_password>"
+    ```
+
+3. Update `generatorConfig.yaml` with the correct settings.
````
The file will come in a later PR?

samuel-duhaime replied:
Yes, it's coming really soon.

tahini commented on generator/generator.md:
```diff
@@ -0,0 +1,246 @@
+| [inputType](#input) | Type of input for the question                          | Number                                                                        | string               |
+| active              | Widget activation status                                | TRUE                                                                          | boolean              |
+| section             | Section to which the question belongs                   | home                                                                          | string               |
+| path                | Path to the data point for the question                 | home.Q2_nbPersonnesResidenceTravailPreciser                                   | string               |
```
Path in the responses object

tahini commented on generator/generator.md:
```diff
@@ -0,0 +1,246 @@
+|-----------------------------|----------------------------------------|-------------------------------|-------------------------|
+| conditionalName             | Name of the conditional                | siPermisConduireConditional   | string                  |
+| logicalOperator             | Logical operator (optional)            | &&                            | && or \|\|              |
+| path                        | Path to the data point for comparison  | [${relativePath}](#rel).age   | string                  |
```
I don't understand the example here. Also "in the responses object" instead of "data point"

tahini replied:
I don't understand the example here. Also "in the responses object" instead of "data point"

tahini commented on generator/generator.md:
```diff
@@ -0,0 +1,246 @@
+
+> <span id="rel">**Note:**</span> `${relativePath}` in `path` is used to obtain the relative path within the same group, facilitating the reference to data points that share a common parent or group with the current data point being processed in the TypeScript code for generating conditionals.
+
+### Conditionals TypeScript
```
Here you should put an example csv data with what is required to generate the code below, with exact fields. That will be more helpful for the documentation readers

samuel-duhaime replied:
Great idea.

tahini commented on generator/generator.md:
```diff
@@ -0,0 +1,246 @@
+| value                        | Unique value for the choice                            | oui            | string or number |
+| fr                           | French label for the choice                            | Oui            | string or number |
+| en                           | English label for the choice                           | Yes            | string or number |
+| [spreadChoicesName](#spread) | Spreading another choicesName (optional)               |                | string?          |
```
I'm not sure I understand the syntax here: `[spreadChoicesName](#spread)`

samuel-duhaime replied:
This allows it to create a link with the note, to make it easier for the user when reading on GitHub.

tahini commented on generator/generator.md:
```diff
@@ -0,0 +1,246 @@
+
+| Field       | Description          | Example                                                           | Type   |
+|-------------|----------------------|-------------------------------------------------------------------|--------|
+| path        | Path to the question | introduction.Q1_quelleOrganisation                                | string |
```
For the labels, you can look at the fillLocales script, which is already in Evolution, and re-use its logic. It allows to add contexts to the labels, like _cati for telephone interviews vs web, or _one for first person vs nickname formulations, with column headers of the type [lang][_context] (eg. fr, fr_cati, fr_one, fr_cati_one) for 4 different labels in french depending on context.

samuel-duhaime replied:
Yes, the generateLibelles.py is pretty much the same file as fillLocales, but I did make some changes to make it work with all the other Generator scripts.

samuel-duhaime replied:
I will add this to the todo list.

davidmurray commented:
Cool @samuel-duhaime  !! 😄 
J'imagine que ça a déjà été discuté, mais pourquoi est-ce qu'on a un script python qui lit le document Excel et qui génère le code typescript? Pourquoi pas avoir du code typescript qui va simplement lire le document Excel directement et générer les widgets à partir de ça?

tahini commented:
> Cool @samuel-duhaime !! 😄 J'imagine que ça a déjà été discuté, mais pourquoi est-ce qu'on a un script python qui lit le document Excel et qui génère le code typescript? Pourquoi pas avoir du code typescript qui va simplement lire le document Excel directement et générer les widgets à partir de ça?

@davidmurray On vise du data-driven complet à long terme. Mais on fait ça évolutif pour Evolution! :D Ici donc, ce sont des scripts python qui vont être roulés offline pour générer du typescript, qui lui sera roulé pour l'enquête.
À court/moyen terme, les scripts python pourraient même être roulés dans le Dockerfile/sur le serveur (ie pas dans git) pour générer le typescript, mais ça restera le typscript qui sera exécuté.
C'est surtout pour le client en fait, les condtionnelles/validations/labels sont des fonctions qui roulent sur le client, qu'on en peut pas sérialiser, et no way que le client va lire un fichier excel de l'enquête!
À long terme, il faudrait que le serveur puisse envoyer l'enquête en entier au client, sérialisée, mais on a un bon bout de chemin à faire avant d'être rendu là! baby steps!

tahini commented:
@davidmurray Merci de poser la question, ça permet de bien clarifier le pourquoi et le documenter en même temps ;-)

davidmurray commented:
> > Cool @samuel-duhaime !! 😄 J'imagine que ça a déjà été discuté, mais pourquoi est-ce qu'on a un script python qui lit le document Excel et qui génère le code typescript? Pourquoi pas avoir du code typescript qui va simplement lire le document Excel directement et générer les widgets à partir de ça?
> 
> @davidmurray On vise du data-driven complet à long terme. Mais on fait ça évolutif pour Evolution! :D Ici donc, ce sont des scripts python qui vont être roulés offline pour générer du typescript, qui lui sera roulé pour l'enquête. À court/moyen terme, les scripts python pourraient même être roulés dans le Dockerfile/sur le serveur (ie pas dans git) pour générer le typescript, mais ça restera le typscript qui sera exécuté. C'est surtout pour le client en fait, les condtionnelles/validations/labels sont des fonctions qui roulent sur le client, qu'on en peut pas sérialiser, et no way que le client va lire un fichier excel de l'enquête! À long terme, il faudrait que le serveur puisse envoyer l'enquête en entier au client, sérialisée, mais on a un bon bout de chemin à faire avant d'être rendu là! baby steps!

Ah!
Je comprends beaucoup mieux maintenant! Merci  pour l’explication détaillée 😊 Il reste effectivement beaucoup de travail à faire pour de rendre à l’étape « long terme » décrite, mais ceci me semble être un excellent début!

samuel-duhaime commented:
@davidmurray Oui en effet, initialement j'avais créé les scripts avec TypeScript, mais en cours de développement, je me suis rendu compte que Python était plus approprié pour le développement de script, je trouvais que les quelques librairies que je me suis servies étaient mieux documentées et faites avec Python surtout pour lire les données directement avec Excel. Aussi, étant une personne curieuse, j'ai voulu apprendre un nouveau langage informatique.

davidmurray commented:
oui Python est très bien pour ce genre de truc! 🙂

samuel-duhaime commented:
@tahini I made the requested adjustments. I think we can accept this review. 

- The documentation is entirely in English
- The examples are clearer
- There are new tables for all csv data example
- To start the project: you can copy a demonstration Excel file (it will be available soon)
- I added a few sections to better explain the project
- I have different instructions for Linux and Windows users
- I added a TODO for the libelles

tahini reviewed (approved):
We can accept, but this will have to be revisited at the very end, when all scripts have been accepted

samuel-duhaime commented on generator/generator.md:
```diff
@@ -75,172 +121,219 @@ In your survey logic, widgets define the various elements that make up your surv
+export const end_email: inputTypes.InputString = {
+    ...defaultInputBase.inputStringBase,
+    path: 'end.email',
+    label: (t: TFunction) => t('end:end.email'),
```
Great idea!
